Source/OCMock/OCMockObject.m

```diff
@@ -261,50 +284,75 @@ - (void)forwardInvocation:(NSInvocation *)anInvocation
     }
     @catch(NSException *e)
     {
-        [exceptions addObject:e];
+        @synchronized(exceptions)
+        {
+            [exceptions addObject:e];
+        }
         [e raise];
     }
 }
 
 - (BOOL)handleInvocation:(NSInvocation *)anInvocation
 {
-    [invocations addObject:anInvocation];
-
+    @synchronized(invocations)
+    {
+        // When the method has a char* argument we do not retain the arguments. This makes it possible
+        // to match char* args literally and with anyPointer. Not retaining the argument means that
+        // in these cases tests that use their own autorelease pools may fail unexpectedly.
+        if(![anInvocation hasCharPointerArgument])
+            [anInvocation retainArguments];
+        [invocations addObject:anInvocation];
+    }
+    
```

Comment:
The arguments are retained as long as **none** of them happen to be a char *.  Which as the comment I copied says, doesn't really work.  But there's too much matching code elsewhere that breaks if you do retain them.  So in the interest of consistency I went with the same system of not retaining char * arguments.

I don't think the invocation is ever actually invoked, it's just kept for matching, mostly in the name of OCMVerify.  But that's an interesting idea, a nil target could be considered matching for self.  I'll see how tricky that is to implement, it seems like the best of both worlds there.